client-mr/src/main/java/org/apache/hadoop/mapred/RssMapOutputCollector.java

```diff
@@ -0,0 +1,204 @@
+
+    int heartBeatThreadNum = jobConf.getInt(RssMRConfig.RSS_CLIENT_HEARTBEAT_THREAD_NUM,
+        RssMRConfig.RSS_CLIENT_HEARTBEAT_THREAD_NUM_DEFAULT_VALUE);
+    int retryMax = jobConf.getInt(RssMRConfig.RSS_CLIENT_RETRY_MAX,
+        RssMRConfig.RSS_CLIENT_RETRY_MAX_DEFAULT_VALUE);
+    long retryIntervalMax = jobConf.getLong(RssMRConfig.RSS_CLIENT_RETRY_INTERVAL_MAX,
+        RssMRConfig.RSS_CLIENT_RETRY_INTERVAL_MAX_DEFAULT_VALUE);
+    String clientType = jobConf.get(RssMRConfig.RSS_CLIENT_TYPE,
+        RssMRConfig.RSS_CLIENT_TYPE_DEFAULT_VALUE);
+
+    int replicaWrite = jobConf.getInt(RssMRConfig.RSS_DATA_REPLICA_WRITE,
+        RssMRConfig.RSS_DATA_REPLICA_WRITE_DEFAULT_VALUE);
+    int replicaRead = jobConf.getInt(RssMRConfig.RSS_DATA_REPLICA_READ,
+        RssMRConfig.RSS_DATA_REPLICA_READ_DEFAULT_VALUE);
+    int replica = jobConf.getInt(RssMRConfig.RSS_DATA_REPLICA,
+        RssMRConfig.RSS_DATA_REPLICA_DEFAULT_VALUE);
+    ShuffleWriteClient client = ShuffleClientFactory
+        .getInstance()
+        .createShuffleWriteClient(clientType, retryMax, retryIntervalMax,
+            heartBeatThreadNum, replica, replicaWrite, replicaRead);
+
+    long sendCheckInterval = jobConf.getLong(RssMRConfig.RSS_CLIENT_SEND_CHECK_INTERVAL_MS,
+        RssMRConfig.RSS_CLIENT_DEFAULT_SEND_CHECK_INTERVAL_MS);
+    long sendCheckTimeout = jobConf.getLong(RssMRConfig.RSS_CLIENT_SEND_CHECK_TIMEOUT_MS,
+        RssMRConfig.RSS_CLIENT_DEFAULT_CHECK_TIMEOUT_MS);
+    int bitmapSplitNum = jobConf.getInt(RssMRConfig.RSS_CLIENT_BITMAP_NUM,
+        RssMRConfig.RSS_CLIENT_DEFAULT_BITMAP_NUM);
+    int numMaps = jobConf.getNumMapTasks();
+    String storageType = jobConf.get(RssMRConfig.RSS_STORAGE_TYPE);
+    if (StringUtils.isEmpty(storageType)) {
+      throw new RssException("storage type mustn't be empty");
+    }
+
+    Map<Integer, List<ShuffleServerInfo>> partitionToServers = Maps.newHashMap();
+    for (int i = 0; i < partitions; i++) {
+      String servers = jobConf.get(RssMRConfig.RSS_ASSIGNMENT_PREFIX + i);
+      if (StringUtils.isEmpty(servers)) {
+        throw new RssException("assign partition " + i + " shouldn't be empty");
+      }
+      String[] splitServers = servers.split(",");
+      List<ShuffleServerInfo> assignServers = Lists.newArrayList();
+      for (String splitServer : splitServers) {
+        String[] serverInfo = splitServer.split(":");
+        if (serverInfo.length != 2) {
+          throw new RssException("partition " + i + " server info isn't right");
+        }
+        ShuffleServerInfo sever = new ShuffleServerInfo(StringUtils.join(serverInfo, "-"),
+            serverInfo[0], Integer.parseInt(serverInfo[1]));
+        assignServers.add(sever);
+      }
+      partitionToServers.put(i, assignServers);
+    }
+    SerializationFactory serializationFactory = new SerializationFactory(jobConf);
+    long maxSegmentSize = jobConf.getLong(RssMRConfig.RSS_CLIENT_MAX_SEGMENT_SIZE,
```

Comment:
maxSegmentSize = max(size of current record, size in config)

Reply:
This is init function, we don't know the size of current record.

Reply:
when create wrappedBuffer size = max(size of current record, size in config)

Reply:
What's the advantages?

Reply:
I use fix length buffer to avoid memory fragments.

Reply:
here is case in spark, the record size > size in config, and the data can't be stored.
job failed.

Reply:
> here is case in spark, the record size > size in config, and the data can't be stored. job failed.

In mr, the record will be split to multiple buffers.